Added basic hover overlay for login options.

Naltharial opened the pull request:
Added a very basic overlay to the Login link in the header. Connected to the BrowserID changes, but sufficiently distinct that they should probably be evaluated independantly.

Mostly just added jQuery handlers to display an overlay when hovering over Login and a bisic template in account/ajax/login.html, because we don't want the whole site to display, just the content.

Commit:
Added basic hover overlay for login options.

mitar commented:
I am not sure why is Ajax needed here? Why do you have to load login options through it?

Naltharial commented:
Well, we don't entirely need it just for this overlay, but then you would have to add ajax/login.html template into the base template and hide the div. I think that would reduce "cleanness" of templating. Maybe not as much with just this one, but if any other overlays are added later, you would need to manually include every overlay template and hide all the `<div>`s. So I think Ajax is the cleanest version to be able to add overlays later or change existing ones without having to change base frontend files.

mitar commented:
> but then you would have to add ajax/login.html template into the base template and hide the div.

I think this is completely normal and usual way to do it. Exactly because of this you split templates into multiple files, so that you can include the same file at different places. And then you wrap that into div, so that you can style them differently, if needed.

> but if any other overlays are added later, you would need to manually include every overlay template and hide all the divs. 

So? Then you just define one big div which includes all hidden layers.

Not for our case, when site is very dynamic, but this is in general better for browsers with bad JavaScript support or no CSS. It still displays the menu. With Ajax, you would lose that content. As I said, this is not really important in our case, as everything is dynamic. But just as a general argument why is nothing wrong with thi.

Ajax version has a delay and unnecessary complexity. Templates were made exactly for this type of situations, where you want to use same output on multiple locations. To deduplicate HTML you would have to do in, for example, static HTML files.

So no, Ajax is completely unneeded here. It does not solve the problem you are describing. Because other approaches exist exactly for this use case.

mitar commented:
Please update from main repository.

Commit:
Upstream merge.

Commit:
Preliminary popup tags.

Commit:
Popups sub-application.

Commit:
Merge remote-tracking branch 'upstream/master' into Asynch

Commit:
Fixed view implementation, added external context.

Naltharial commented:
Well, since Ajax was inappropriate, I implemented it with templates. Since the popups might sometimes need a more complex logic, I've implemented a custom template tag to allow including views as popups. Since external context is pushed through, we can use `sekizai` to add to base blocks from popups.
All that is needed in base template is:

```
<% popup <name of view> %>
```

and the application creates the view, renders it and returns the content.

It might be a bit complex for the task, but I wanted to experiment with some of the more powerful Django tools so that we can implement almost anything as popups if needed.

mitar commented:
This was done much more simpler with #270. This is an interesting approach, but also quite broken. Will comment more.

mitar commented:
In general, you are mixing flows. First we have views, which prepare context, and then template rendering. And you in template rendering again go to views. This is not a clean design. What you can do is tools provided in templates to break them in parts (inheritance or inclusion). What was done in #270.

So in general mixing control logic with template logic is bad. It is much harder to debug. So we should not have magical template tags, for example. It is possible to do everything in template tags, but then we have PHP and not Django. :-)

mitar commented on piplmesh/frontend/static/piplmesh/js/common.js:
```diff
@@ -1,4 +1,22 @@
+		$("#overlay").fadeIn(150);
+		
+		// Close on the X icon; .live() monitors changes to the DOM
+		$("#overlay .close").live('click', function(){
```
`.on` is currently recommended.

mitar commented on piplmesh/popups/templatetags/popup.py:
```diff
@@ -0,0 +1,29 @@
+
+register = template.Library()
+
+class PopupNode(template.base.Node):
```
You could use [simple_tag](https://docs.djangoproject.com/en/dev/howto/custom-template-tags/#simple-tags) decorator. For this template tag there is no need to do custom `Node` class.

mitar commented on piplmesh/popups/templatetags/popup.py:
```diff
@@ -0,0 +1,29 @@
+    bits = token.split_contents()[1:]
+    
+    if hasattr(views, bits[0]):
+        return PopupNode(getattr(views, bits[0]))
```
You hardcode here views. If you would really want to do this, it would be much better to require whole Python path to the view where you would then here import the module and use that view.

mitar commented on piplmesh/popups/views.py:
```diff
@@ -0,0 +1,29 @@
+from django.contrib.auth import forms as auth_forms
+from django.views.generic import edit as edit_views
+
+class ExternalContextMixin():
```
Really interesting way to put context back into the view. Really crazy ideas. :-) (In a positive way. Sometimes it is interesting to see such solutions because I would never imagine them because they are so out of the "right way". But I can see that this could maybe be of use someday. But for current problem this is really a hammer to simple problem of including template.)

mitar commented on piplmesh/urls.py:
```diff
@@ -40,6 +40,8 @@
     url(r'^login/$', 'django.contrib.auth.views.login', {'template_name': 'user/login.html'}, name='login'),
     url(r'^logout/$', account_views.logout, name='logout'),
+    
+    url(r'^ajax/login/$', 'django.contrib.auth.views.login', {'template_name': 'ajax/login.html'}, name='ajax/login'),
```
This should be cleaned.

mitar commented on piplmesh/popups/templatetags/popup.py:
```diff
@@ -0,0 +1,29 @@
+
+    def render(self, context):
+        view_function = self.popup.as_view(context=context)
+        template_response = view_function(context['request'])
```
See. This is the main problem of your approach. View functions are wrapped in middleware. You don't run them here. So there is some pre and post processing possible. What if author of the view function counts on that?

mitar commented:
The problem is that it works also only with class-based views.

mitar commented:
Anyway, an interesting approach. And some ingenious solutions. You made this yourself? Because this is some hard stuff. If you did it yourself, I will give you some points for it.

But sorry, this will not go in.

Naltharial commented:
Heh, it's ok, I knew it was something of an experiment to begin with. At least I learned quite a bit about views in the process.

Yes, I did it myself. I thought I was stuck at one point
http://stackoverflow.com/questions/12262654/django-view-not-updating-context-properly/
but then I figured out that me injecting views in the middle of template processing was exactly the problem. So I injected the context into it as well, and it worked. Well, for this current view at least, unfortunately a lot of django views have differing implementations of `get_context_data()`.
